### DISCUSSION TITLE
Dynamically calculate normals/tangents

### DISCUSSION THREAD

**Discussion by paulosensorydata** [google/filament#4407] [Q&A] [answered]
Hello everyone. I have been trying to convert my SceneKit development into Filament (at least for Android)

I create dynamically a geometry with 400 vertices -> 2166 indices
Successfully right now. I am also using the an Orbit Manipulator correctly.

I am now trying to add lightning now but I realise I need to add normals first. From what I saw, I also need tangents and bitangents to be added (and calculated)

```
val vertexPositionSize = 3 * floatSize
val vertexColorSize = 1 * intSize

data class Vertex(val x: Float, val y: Float, val z: Float)
data class Color(val color: Int)

fun ByteBuffer.put(v: Vertex): ByteBuffer {
    putFloat(v.x)
    putFloat(v.y)
    putFloat(v.z)
    return this
}

fun ByteBuffer.putColor(v: Color): ByteBuffer {
    putInt(v.color)
    return this
}

val vertexData = ByteBuffer.allocate(vertexCount * vertexPositionSize)
    .order(ByteOrder.nativeOrder())

val vertexColor = ByteBuffer.allocate(vertexCount * vertexColorSize)
    .order(ByteOrder.nativeOrder())

values.forEachIndexed { x, row ->
    row.forEachIndexed { y, value ->
        vertexData.put(Vertex(x.toFloat() * 20, y.toFloat() * 20, value.toFloat()))
        vertexColor.putColor(Color(getRgbColor(value.toFloat() / 400)))
    }
}

vertexData.flip()
vertexColor.flip()
```

Is there a way to dynamically calculate normals and tangents in this case?

**Comment by romainguy** [comment, marked as answer]
To do this automatically you can use the `SurfaceOrientation` class. It offers several methods that will be more or less precise based on:

- normals only (not recommended)
- normals + tangents (sign of W determines bitangent orientation)
- normals + uvs + positions + indices
- positions + indices

In addition if you won't use normal mapping or anisotropy, you can just make up the tangents/bitangents (set one to +X and the other one to +Y always for instance).

**Reply by paulosensorydata** [reply]
Do I need to call this once for the whole geometry like:

```
val surfaceOrientation = SurfaceOrientation
            .Builder()
            .vertexCount(400)
            .positions(vertexData)
            .triangleCount(722)
            .triangles_uint16(indexData)
            .build()
```

Or once per triangle?
I believe the next step is to call `surfaceOrientation.getQuatsAsFloat()`and use the output somewhere, right?
Also: I am unsure what should be passed as buffer on this method.

Sorry for all these questions and thanks for all your answers @romainguy

**Reply by romainguy** [reply]
You do this for the entire geometry, doing this once per triangle would be way too expensive

> Also: I am unsure what should be passed as buffer on this method.

You need to pass a `Buffer` you allocated so we can store the results somewhere. That buffer will then be used when you create your `VertexBuffer` to define your tangents data.
